### PR TITLE
add new way of iterating over mobs. Hopefully fixes the proc crash

### PR DESCRIPTION
**Pull request by cizra**

### CODE REVIEW

File: code/code/sys/CharacterList.cc [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+extern TBeing *character_list;
+
+TCharacterList::iterator::iterator(TBeing* d)
+  : d(d)
+{}
+
+TCharacterList::iterator TCharacterList::iterator::operator++()
+{
+  d = d->next;
```
Comment (wershlak) [inline code/code/sys/CharacterList.cc:13]: Won't we see a crash here in the case the being that d points to is deleted?

**Comment by wershlak** [comment]
My thought to how to fix this tmp_ch issue was that in the TBeing destructor we need to move the tmp_ch pointer in the case that is the thing we are deleting.

```
if this == tmp_ch
  tmp_ch = tmp_ch->next
```
Or some such thing.

At least that seems the most straightforward. It doesn't really feel good though

File: code/code/sys/CharacterList.cc (hunk 1 above)
Comment (cizra) [reply code/code/sys/CharacterList.cc:13]: We would, but modifying the loop range while looping is a really, really bad idea anyway. I hope we'll never do that again.

**Comment by cizra** [comment]
I agree, it'd be a simple hack - reaching to the other end of the code, poke the private variable inside an unrelated class... thankfully it's a private variable, which exposes the uncleanliness of such a movement.

**Comment by wershlak** [comment]
Alright cool. Glad that makes sense. 

You have this running right? Are you satisfied this fixes it?

**Review by wershlak — APPROVED** [review]

**Comment by cizra** [comment]
It's currently running in prod, yeah.
Absolutely not sure! :D I can only hope it works.

**Pull request merged**